File: frontend/lib/models/recycling_point.dart
class RecyclingPoint {
  final String id;
  final String name;
  final List<String> type;
  final String address;
  final String city;
  final String country;
  final PointCoordinates coordinates;
  final String? workingHours;
  final String? phone;
  final String? website;
  final String? description;
  final List<String> photos;
  final String? addedById;
  final String? addedByName;
  final bool verified;
  final double rating;
  final List<PointReview> reviews;
  final DateTime createdAt;

  RecyclingPoint({
    required this.id,
    required this.name,
    required this.type,
    required this.address,
    required this.city,
    required this.country,
    required this.coordinates,
    this.workingHours,
    this.phone,
    this.website,
    this.description,
    required this.photos,
    this.addedById,
    this.addedByName,
    required this.verified,
    required this.rating,
    required this.reviews,
    required this.createdAt,
  });

  factory RecyclingPoint.fromJson(Map<String, dynamic> json) {
    return RecyclingPoint(
      id: json['_id'],
      name: json['name'],
      type: List<String>.from(json['type'] ?? []),
      address: json['address'],
      city: json['city'],
      country: json['country'] ?? 'Россия',
      coordinates: PointCoordinates.fromJson(json['coordinates']),
      workingHours: json['workingHours'],
      phone: json['phone'],
      website: json['website'],
      description: json['description'],
      photos: List<String>.from(json['photos'] ?? []),
      addedById: json['addedBy'] is String 
          ? json['addedBy'] 
          : json['addedBy']?['_id'],
      addedByName: json['addedBy'] is Map 
          ? json['addedBy']['username'] 
          : null,
      verified: json['verified'] ?? false,
      rating: (json['rating'] ?? 0).toDouble(),
      reviews: (json['reviews'] as List?)
          ?.map((r) => PointReview.fromJson(r))
          .toList() ?? [],
      createdAt: DateTime.parse(json['createdAt']),
    );
  }

  int get reviewCount => reviews.length;
  String get typeString => type.join(', ');
}

class PointCoordinates {
  final String type;
  final List<double> coordinates; // [longitude, latitude]

  PointCoordinates({
    required this.type,
    required this.coordinates,
  });

  factory PointCoordinates.fromJson(Map<String, dynamic> json) {
    return PointCoordinates(
      type: json['type'] ?? 'Point',
      coordinates: List<double>.from(json['coordinates']),
    );
  }

  double get longitude => coordinates[0];
  double get latitude => coordinates[1];
}

class PointReview {
  final String userId;
  final String? username;
  final int rating;
  final String? comment;
  final DateTime createdAt;

  PointReview({
    required this.userId,
    this.username,
    required this.rating,
    this.comment,
    required this.createdAt,
  });

  factory PointReview.fromJson(Map<String, dynamic> json) {
    return PointReview(
      userId: json['user'] is String ? json['user'] : json['user']['_id'],
      username: json['user'] is Map ? json['user']['username'] : null,
      rating: json['rating'],
      comment: json['comment'],
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
}

File: backend/models/RecyclingPoint.js
const mongoose = require('mongoose');

const recyclingPointSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Название пункта обязательно'],
    trim: true
  },
  type: [{
    type: String,
    enum: ['plastic', 'glass', 'paper', 'metal', 'electronics', 'batteries', 'clothes', 'hazardous', 'organic'],
    required: true
  }],
  address: {
    type: String,
    required: [true, 'Адрес обязателен']
  },
  city: {
    type: String,
    required: [true, 'Город обязателен']
  },
  country: {
    type: String,
    default: 'Россия'
  },
  coordinates: {
    type: {
      type: String,
      enum: ['Point'],
      default: 'Point'
    },
    coordinates: {
      type: [Number], // [longitude, latitude]
      required: [true, 'Координаты обязательны']
    }
  },
  workingHours: String,
  phone: String,
  website: String,
  description: String,
  photos: [String], // URLs фотографий
  addedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  verified: {
    type: Boolean,
    default: false
  },
  rating: {
    type: Number,
    default: 0,
    min: 0,
    max: 5
  },
  reviews: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    rating: {
      type: Number,
      required: true,
      min: 1,
      max: 5
    },
    comment: String,
    createdAt: {
      type: Date,
      default: Date.now
    }
  }]
}, {
  timestamps: true
});

// Geo-индекс для поиска по координатам
recyclingPointSchema.index({ coordinates: '2dsphere' });
recyclingPointSchema.index({ city: 1 });
recyclingPointSchema.index({ type: 1 });
recyclingPointSchema.index({ verified: 1 });

// Виртуальное поле для количества отзывов
recyclingPointSchema.virtual('reviewCount').get(function() {
  return this.reviews.length;
});

// Метод для пересчета среднего рейтинга
recyclingPointSchema.methods.calculateAverageRating = function() {
  if (this.reviews.length === 0) {
    this.rating = 0;
  } else {
    const sum = this.reviews.reduce((acc, review) => acc + review.rating, 0);
    this.rating = (sum / this.reviews.length).toFixed(1);
  }
  return this.save();
};

// Метод для добавления отзыва
recyclingPointSchema.methods.addReview = function(userId, rating, comment) {
  // Проверяем, не оставлял ли пользователь уже отзыв
  const existingReview = this.reviews.find(r => r.user.toString() === userId.toString());
  
  if (existingReview) {
    // Обновляем существующий отзыв
    existingReview.rating = rating;
    existingReview.comment = comment;
  } else {
    // Добавляем новый отзыв
    this.reviews.push({ user: userId, rating, comment });
  }
  
  return this.calculateAverageRating();
};

// Статический метод для поиска ближайших пунктов
recyclingPointSchema.statics.findNearby = function(longitude, latitude, maxDistance = 5000, type = null) {
  const query = {
    coordinates: {
      $near: {
        $geometry: {
          type: 'Point',
          coordinates: [longitude, latitude]
        },
        $maxDistance: maxDistance // в метрах
      }
    },
    verified: true
  };

  if (type) {
    query.type = { $in: Array.isArray(type) ? type : [type] };
  }

  return this.find(query);
};

module.exports = mongoose.model('RecyclingPoint', recyclingPointSchema);

